<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Venndir with Gene Expression Data • venndir</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Venndir with Gene Expression Data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">venndir</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.57.900</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/venndir_gene_expression.html">Venndir with Gene Expression Data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jmw86069/venndir/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Venndir with Gene Expression Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jmw86069/venndir/blob/HEAD/vignettes/venndir_gene_expression.Rmd" class="external-link"><code>vignettes/venndir_gene_expression.Rmd</code></a></small>
      <div class="d-none name"><code>venndir_gene_expression.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jmw86069.github.io/venndir">venndir</a></span><span class="op">)</span></span></code></pre></div>
<p>Venndir was driven by gene expression analysis, which contains
up/down directionality. Here are some examples showing how gene
expression data can be used in <code>venndir</code>.</p>
<p>The examples include <code>limma</code> and <code>DESeq2</code> as
two common example workflows.</p>
<div class="section level2">
<h2 id="generic-example">Generic example<a class="anchor" aria-label="anchor" href="#generic-example"></a>
</h2>
<p>The <code><a href="../reference/make_venn_test.html">venndir::make_venn_test()</a></code> function is useful to
generate test data, just to create a Venn diagram.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">setlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_venn_test.html">make_venn_test</a></span><span class="op">(</span><span class="fl">2500</span>, <span class="fl">3</span>,</span>
<span>  sizes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">500</span>, <span class="fl">200</span><span class="op">)</span>,</span>
<span>  do_signed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">setlist</span>, overlap_type<span class="op">=</span><span class="st">"agreement"</span><span class="op">)</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/generic_1-1.png" alt="Venndir default example showing signed counts, usingoverlap type 'agreement'." width="960"></p>
</div>
<div class="section level2">
<h2 id="venndir-with-limma-results">Venndir with limma results<a class="anchor" aria-label="anchor" href="#venndir-with-limma-results"></a>
</h2>
<div class="section level3">
<h3 id="illumina-case-study">Illumina Case Study<a class="anchor" aria-label="anchor" href="#illumina-case-study"></a>
</h3>
<p>The <code>limma</code> package provides an extensive set of gene
expression data analysis tools. The steps below reproduce the
<code>Limma Users Guide</code> example <code>"Illumina Use Case"</code>
in Section 17.3. This example requires data available at <a href="https://bioinf.wehi.edu.au/marray/IlluminaCaseStudy" class="external-link uri">https://bioinf.wehi.edu.au/marray/IlluminaCaseStudy</a>
copied to the current working directory.</p>
<p>For the sake of this example, the steps assume the data is located in
a folder <code>[HOME]/IlluminaCaseStudy</code>, where
<code>[HOME]</code> is the user home directory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">illuminadir</span> <span class="op">&lt;-</span> <span class="st">"~/IlluminaCaseStudy"</span>;</span>
<span><span class="va">do_limma</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>;</span>
<span><span class="va">do_deseq2</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>;</span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="va">illuminadir</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">do_limma</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span>;</span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">do_deseq2</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span>;</span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/readTargets.html" class="external-link">readTargets</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">illuminadir</span><span class="op">)</span></span>
<span>   <span class="co"># import Illumina data</span></span>
<span>   <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/read.ilmn.html" class="external-link">read.ilmn</a></span><span class="op">(</span>files<span class="op">=</span><span class="st">"probe profile.txt.gz"</span>,</span>
<span>      ctrlfiles<span class="op">=</span><span class="st">"control probe profile.txt.gz"</span>,</span>
<span>      other.columns<span class="op">=</span><span class="st">"Detection"</span>,</span>
<span>      path<span class="op">=</span><span class="va">illuminadir</span><span class="op">)</span></span>
<span>   <span class="co"># define expressed probes</span></span>
<span>   <span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/nec.html" class="external-link">neqc</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>   <span class="va">expressed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">Detection</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span>;</span>
<span>   <span class="co"># subset for expressed probes</span></span>
<span>   <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">expressed</span>,<span class="op">]</span></span>
<span>   <span class="co"># calculate within-patient correlations</span></span>
<span>   <span class="va">ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">targets</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span></span>
<span>   <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">ct</span><span class="op">)</span>;</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Factor-class.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">ct</span><span class="op">)</span>;</span>
<span>   <span class="va">dupcor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">duplicateCorrelation</a></span><span class="op">(</span><span class="va">y</span>,</span>
<span>      <span class="va">design</span>,</span>
<span>      block<span class="op">=</span><span class="va">targets</span><span class="op">$</span><span class="va">Donor</span><span class="op">)</span>;</span>
<span>   <span class="va">dupcor</span><span class="op">$</span><span class="va">consensus.correlation</span>;</span>
<span>   </span>
<span>   <span class="co"># fit linear model, paired by patient, with correlations</span></span>
<span>   <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">y</span>,</span>
<span>      <span class="va">design</span>,</span>
<span>      block<span class="op">=</span><span class="va">targets</span><span class="op">$</span><span class="va">Donor</span>,</span>
<span>      correlation<span class="op">=</span><span class="va">dupcor</span><span class="op">$</span><span class="va">consensus.correlation</span><span class="op">)</span>;</span>
<span>   <span class="co"># define contrasts</span></span>
<span>   <span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span><span class="va">ML</span><span class="op">-</span><span class="va">MS</span>, </span>
<span>      <span class="va">LP</span><span class="op">-</span><span class="va">MS</span>, </span>
<span>      <span class="va">ML</span><span class="op">-</span><span class="va">LP</span>, </span>
<span>      levels<span class="op">=</span><span class="va">design</span><span class="op">)</span></span>
<span>   </span>
<span>   <span class="co"># fit contrasts</span></span>
<span>   <span class="va">fit2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">contrasts</span><span class="op">)</span>;</span>
<span>   </span>
<span>   <span class="co"># eBayes</span></span>
<span>   <span class="va">fit2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit2</span>, trend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Reading file ~/IlluminaCaseStudy/probe profile.txt.gz ... ...</span></span>
<span><span class="co">#&gt; Reading file ~/IlluminaCaseStudy/control probe profile.txt.gz ... ...</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="lmfit-or-ebayes-to-incidence-matrix">lmFit() or eBayes() to incidence matrix<a class="anchor" aria-label="anchor" href="#lmfit-or-ebayes-to-incidence-matrix"></a>
</h3>
<p>The function <code><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">limma::decideTests()</a></code> will convert the
output <code>MArrayLM</code> object (from <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">limma::lmFit()</a></code> or
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">limma::eBayes()</a></code>), first applying thresholds for adjusted
P-value and log2 fold change.</p>
<p>The output has class <code>TestResults</code> and is a signed
incidence matrix! The values are already: <code>c(-1, 0, 1)</code>,
which is perfect!</p>
<p>Take a look at the output:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="co"># decideTests() creates a signed incidence matrix</span></span>
<span>   <span class="va">fit2decide</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">fit2</span>,</span>
<span>      p.value<span class="op">=</span><span class="fl">0.01</span>, lfc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>      method<span class="op">=</span><span class="st">"global"</span><span class="op">)</span>;</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">fit2decide</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; TestResults matrix</span></span>
<span><span class="co">#&gt;               Contrasts</span></span>
<span><span class="co">#&gt;                ML - MS LP - MS ML - LP</span></span>
<span><span class="co">#&gt;   ILMN_2055271       0       0       0</span></span>
<span><span class="co">#&gt;   ILMN_1653355       0       0       0</span></span>
<span><span class="co">#&gt;   ILMN_1787689       0       0       0</span></span>
<span><span class="co">#&gt;   ILMN_1745607      -1      -1       0</span></span>
<span><span class="co">#&gt;   ILMN_1735045       0       0       0</span></span>
<span><span class="co">#&gt; 24686 more rows ...</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="examples-by-input-type">Examples by Input Type<a class="anchor" aria-label="anchor" href="#examples-by-input-type"></a>
</h2>
<div class="section level3">
<h3 id="incidence-matrix">Incidence matrix<a class="anchor" aria-label="anchor" href="#incidence-matrix"></a>
</h3>
<p>A signed incidence matrix can be used directly, it is automatically
converted to a signed <code>list</code>.</p>
<p>For simplicity, we will show only the first two contrasts, using
argument: <code>sets=c(1, 2)</code></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">fit2decide</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_3a-1.png" alt="Venndir using input in the form of a signed incidence matrix, values are -1, 0, or +1." width="960"></p>
</div>
<div class="section level3">
<h3 id="signed-set-list">Signed set list<a class="anchor" aria-label="anchor" href="#signed-set-list"></a>
</h3>
<p>The signed incidence matrix can be converted to a signed set list
using <code><a href="../reference/im_value2list.html">im_value2list()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">limmalist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide</span><span class="op">)</span>;</span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_3-1.png" alt="Venndir using a signed setlist as input, after converting the signed incidence matrix to a signed setlist." width="960"></p>
<p>A few features are noticeable:</p>
<ol style="list-style-type: decimal">
<li><p>The <code>limma</code> contrasts <code>"ML - MS"</code> and
<code>"LP - MS"</code> are shown, both involve a comparison “versus MS”.
It compares the ML differences from MS, to the LP differences from
MS.</p></li>
<li><p>The numbers are statistically significant changes given the
statistical thresholds used by
<code><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">limma::decideTests()</a></code>.</p></li>
<li><p>There is fairly high overlap between these gene lists,
<code>1,582</code> are shared.</p></li>
<li>
<p>All of the shared changes have the same direction of change in
<code>ML-MS</code> compared with <code>LP-MS</code>.</p>
<ul>
<li>
<code>740</code> are up in both comparisons.</li>
<li>
<code>842</code> are down in both comparisons.</li>
</ul>
</li>
</ol>
<blockquote>
<p>Exercise for the reader: Consider using more lenient statistical
thresholds, and test the overlap and directional concordance. Would you
expect more lenient results to have the same concordance, or less
concordance?</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="customizing-the-diagram">Customizing the Diagram<a class="anchor" aria-label="anchor" href="#customizing-the-diagram"></a>
</h2>
<div class="section level3">
<h3 id="proportional-venn-euler">Proportional Venn (Euler)<a class="anchor" aria-label="anchor" href="#proportional-venn-euler"></a>
</h3>
<p>The data can be drawn with proportional circles, which is a Euler
diagram.</p>
<p>For this diagram, we plot <code>sets=c(1, 3)</code> mostly because
these are more visually interesting.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, proportional<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_3p-1.png" alt="Venndir figure with 'proportional=TRUE' which produces a proportional Euler diagram. Signed counts are displayed beside the main overlap counts by default." width="960"></p>
</div>
<div class="section level3">
<h3 id="customizing-the-labels">Customizing the labels<a class="anchor" aria-label="anchor" href="#customizing-the-labels"></a>
</h3>
<p>Sometimes proportional circles have tall-skinny overlap regions, so
you can arrange the signed labels using argument:
<code>template="tall"</code>.</p>
<p>For kicks, we will also enlarge the main count labels using the
second value in <code>font_cex=c(1.2, 1.5, 0.8)</code>. The first value
controls the set label, the second controls the main count label, the
third controls the signed count labels.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, proportional<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>      font_cex<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1.5</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>      template<span class="op">=</span><span class="st">"tall"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p>&lt;img
src=“/Users/jamesward/Projects/Ward/venndir/docs/articles/venndir_gene_expression_files/figure-html/limma_3p_tall-1.png”
alt=“Venndir figure shown as a proportional Euler diagram as before,
except using ‘template=“tall”’ which arranges signed counts below the
main counts.” width=“960” /&gt;</p>
</div>
<div class="section level3">
<h3 id="overlap_typeagreement">overlap_type=“agreement”<a class="anchor" aria-label="anchor" href="#overlap_typeagreement"></a>
</h3>
<p>The argument <code>overlap_type="agreement"</code> will summarize
counts by agreement <code>"="</code>, or disagreement
<code>"X"</code>.</p>
<p>Notice that sets 2 and 3 are predominantly discordant. It may be
confusing, but for this type of analysis, and these comparisons, this
result is expected. The group <code>"LP"</code> is the test group in the
first comparison, and <code>"LP"</code> is the control group in the
second comparison.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, proportional<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>      font_cex<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1.5</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>      template<span class="op">=</span><span class="st">"tall"</span>,</span>
<span>      overlap_type<span class="op">=</span><span class="st">"agreement"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p>&lt;img
src=“/Users/jamesward/Projects/Ward/venndir/docs/articles/venndir_gene_expression_files/figure-html/limma_4-1.png”
alt=“Venndir figure shown as a proportional Euler diagram, with the
option ‘overlap_type=“agreement”’ which summarizes the signed overlaps
as agreeing, or disagreeing in direction.” width=“960” /&gt;</p>
</div>
<div class="section level3">
<h3 id="overlap_typeeach">overlap_type=“each”<a class="anchor" aria-label="anchor" href="#overlap_typeeach"></a>
</h3>
<p>The argument <code>overlap_type="each"</code> will display counts for
each combination of up and down. This option gives the most detail, but
can sometimes be challenging to label clearly.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>      font_cex<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1.5</span>, <span class="fl">0.8</span><span class="op">)</span>,</span>
<span>      template<span class="op">=</span><span class="st">"tall"</span>,</span>
<span>      overlap_type<span class="op">=</span><span class="st">"each"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p>&lt;img
src=“/Users/jamesward/Projects/Ward/venndir/docs/articles/venndir_gene_expression_files/figure-html/limma_5-1.png”
alt=“Venndir figure shown as a proportional Euler diagram, with the
option ‘overlap_type=“each”’ which tabulates the signed counts for each
combination of up and down directions. Again the option
‘template=“tall”’ is used, to arrange signed counts below the main
overlap counts.” width=“960” /&gt;</p>
</div>
</div>
<div class="section level2">
<h2 id="item-labels">Item labels<a class="anchor" aria-label="anchor" href="#item-labels"></a>
</h2>
<p>It is sometimes useful to show item labels inside the Venn diagram.
<strong>However please don’t try it with 3,000+ labels!</strong></p>
<p>For this example, we apply a stringent statistical filter to reduce
the number of hits.</p>
<div class="section level3">
<h3 id="sign-item-label">Sign Item label<a class="anchor" aria-label="anchor" href="#sign-item-label"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="co"># increase stringency of statistical filtering</span></span>
<span>   <span class="va">fit2decide2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">fit2</span>,</span>
<span>      p.value<span class="op">=</span><span class="fl">1e-4</span>,</span>
<span>      lfc<span class="op">=</span><span class="fl">5</span><span class="op">)</span>;</span>
<span>   <span class="co"># increase stringency of statistical filtering</span></span>
<span>   <span class="va">limmalist2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide2</span><span class="op">)</span>;</span>
<span>   <span class="co"># display item labels</span></span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">fit2decide2</span>,</span>
<span>      sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>      show_labels<span class="op">=</span><span class="st">"Ni"</span>,</span>
<span>      item_degrees<span class="op">=</span><span class="fl">2</span>,</span>
<span>      fontfamily<span class="op">=</span><span class="st">"ArialNarrow"</span>,</span>
<span>      show_items<span class="op">=</span><span class="st">"sign item"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_6-1.png" alt="Venndir figure showing item labels filled inside each corresponding overlap region. Labels include the arrow indication direction, beside each item label. Labels are also colored based upon direction. The labels indicate Illumina probe names, since the statistical analysis involved probes, and not specific genes measured by the probes." width="960"></p>
</div>
<div class="section level3">
<h3 id="gene-labels">Gene labels<a class="anchor" aria-label="anchor" href="#gene-labels"></a>
</h3>
<p>This step “gets into the weeds” so to speak, but is more interesting
than printing Illumina probe names. There is a little helper function
<code><a href="../reference/collapse_im.html">collapse_im()</a></code> intended to collapse an incidence matrix by
row groups, which is a fancy way of saying “Make it per-gene”.</p>
<ol style="list-style-type: decimal">
<li>Convert Illumina probe names to Illumina gene symbols.</li>
<li>Summarize the direction per gene with
<code><a href="../reference/collapse_im.html">collapse_im()</a></code>.</li>
</ol>
<p>The method prioritizes non-zero values, so that any evidence of a hit
will be retained, using the majority direction.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="co"># filter for non-zero values</span></span>
<span>   <span class="va">fit2decide2g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">fit2decide2</span>, <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">fit2decide2</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>   <span class="va">match2g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fit2decide2g</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="va">y_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">SYMBOL</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">""</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,</span>
<span>      <span class="va">y</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">SYMBOL</span><span class="op">)</span></span>
<span>   <span class="va">row_genes</span> <span class="op">&lt;-</span> <span class="va">y_symbol</span><span class="op">[</span><span class="va">match2g</span><span class="op">]</span></span>
<span>   <span class="va">fit2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/collapse_im.html">collapse_im</a></span><span class="op">(</span><span class="va">fit2decide2g</span>, <span class="va">row_genes</span><span class="op">)</span></span>
<span>   </span>
<span>   <span class="co"># display item labels</span></span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">fit2gene</span>,</span>
<span>      xyratio<span class="op">=</span><span class="fl">2.2</span>,</span>
<span>      <span class="co"># fontfamily="ArialNarrow",</span></span>
<span>      fontfamily<span class="op">=</span><span class="st">"Avenir Next Condensed"</span>,</span>
<span>      sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>      show_labels<span class="op">=</span><span class="st">"Ni"</span>,</span>
<span>      show_items<span class="op">=</span><span class="st">"sign item"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_6g-1.png" alt="Venndir figure showing item labels which now use the gene symbol, after converting Illumina probe names to the measured genes. Because multiple probes may measure the same gene, the conversion utilizes a helper function: 'im_collapse()'." width="960"></p>
</div>
<div class="section level3">
<h3 id="arrow-labels">Arrow labels<a class="anchor" aria-label="anchor" href="#arrow-labels"></a>
</h3>
<p>When there are many items, sometimes it’s nice to see just the
arrows. It saves space, while also filling the polygon with the color
associated with each direction.</p>
<p>Use argument: <code>show_items="sign"</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">vo6s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">fit2decide2</span>,</span>
<span>      sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>      label_style<span class="op">=</span><span class="st">"shaded box"</span>,</span>
<span>      proportional<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>      show_labels<span class="op">=</span><span class="st">"NcSi"</span>,</span>
<span>      show_items<span class="op">=</span><span class="st">"sign"</span>,</span>
<span>      item_cex<span class="op">=</span><span class="fl">2</span>,</span>
<span>      item_degrees<span class="op">=</span><span class="fl">7</span>,</span>
<span>      max_items<span class="op">=</span><span class="fl">1000</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_6s-1.png" alt="Venndir figure showing a proportional Euler diagram, including item labels in each overlap region. The item labels now only show the directional arrow, not the label itself, in order to indicate the proportion of directions observed. The count labels are placed outside the overlap regions, connected by a line segment." width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="three-way-directional-venn">Three-way directional Venn<a class="anchor" aria-label="anchor" href="#three-way-directional-venn"></a>
</h2>
<p>The examples above purposefully used only two contrasts, because
those two contrasts show very high concordance.</p>
<div class="section level3">
<h3 id="three-group-analysis">Three-group analysis<a class="anchor" aria-label="anchor" href="#three-group-analysis"></a>
</h3>
<p>One common type of analysis involves three experimental groups. For
example:</p>
<ol style="list-style-type: decimal">
<li>Control</li>
<li>Treatment A</li>
<li>Treatment B</li>
</ol>
<p>The typical contrasts are:</p>
<ol style="list-style-type: decimal">
<li>A - Control</li>
<li>B - Control</li>
<li>A - B</li>
</ol>
<p>When creating a three-way Venn diagram, the results are also fairly
straightforward. When the directionality is included, it can be more
confusing, but also potentially much more informative. Note that the
contrast (3) may be written <code>"A - B"</code> or
<code>"B - A"</code>, and the order affects the directionality.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">fit2decide1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">fit2</span>,</span>
<span>      p.value<span class="op">=</span><span class="fl">1e-3</span>,</span>
<span>      lfc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span>;</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">fit2decide1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A - C"</span>,</span>
<span>      <span class="st">"B - C"</span>,</span>
<span>      <span class="st">"A - B"</span><span class="op">)</span></span>
<span>   <span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">fit2decide1</span>,</span>
<span>      sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>      overlap_type<span class="op">=</span><span class="st">"agreement"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p>&lt;img
src=“/Users/jamesward/Projects/Ward/venndir/docs/articles/venndir_gene_expression_files/figure-html/venn_compare_0-1.png”
alt=“Venndir 3-way figure showing counts with
‘overlap_type=“agreement”’. Each set represents the hits from one
statistical contrast, after filtering by P-value and log fold change.”
width=“960” /&gt;</p>
<p>There are a few potential scenarios:</p>
<ol style="list-style-type: decimal">
<li>Treatment A does not affect the same genes as B. Few overlaps.</li>
<li>Treatment A affects similar genes as B, but not with high
concordance.</li>
<li>Treatment A affects similar genes as B, with high concordance.</li>
</ol>
<p>An interesting question is whether A and B are different from each
other. This is contrast (3) above: <code>"A - B"</code></p>
<p>We’ll use the Venndir highlight feature to walk through the
logic.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">v2</span> <span class="op">&lt;-</span> <span class="fu">venndir</span><span class="fu">::</span><span class="fu"><a href="../reference/modify_venndir_overlap.html">highlight_venndir_overlap</a></span><span class="op">(</span><span class="va">v</span>,</span>
<span>      overlap_set<span class="op">=</span><span class="st">"A - B"</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/venn_compare_AB-1.png" alt="Venndir figure shown as in the previous figure, except with one overlap region highlighted using yellow fill color, and a wide orange inner border. The highlighted region represents contrast 'A - B', and the hits which are unique to this comparison. The highlighting helps emphasize this region for discussion." width="960"></p>
<p>These <strong>46</strong> hits are different between A and B, but are
not independently different when comparing A to C, nor B to C. What does
that mean?</p>
<ul>
<li>It suggests that these are moderate treatment effects, because they
aren’t statistically significant when comparing to Control.</li>
<li>It also suggests that the moderate effects are in opposite direction
in A and B compared to Control. (It is possible that Control has higher
variability for these genes, but for now let’s assume variability is
reasonably consistent for all groups.)</li>
</ul>
<p>Taken together, one might expect the number here (46) to be lower
than those for <code>"A - C"</code> or <code>"B - C"</code> simply
because they are <strong>only</strong> hits when comparing A and B
directly. If they were changes also seen in <code>"A - C"</code> they
would appear in the section <strong>214</strong>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">v3</span> <span class="op">&lt;-</span> <span class="fu">venndir</span><span class="fu">::</span><span class="fu"><a href="../reference/modify_venndir_overlap.html">highlight_venndir_overlap</a></span><span class="op">(</span><span class="va">v</span>,</span>
<span>      overlap_set<span class="op">=</span><span class="st">"A - C&amp;A - B"</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v3</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/venn_compare_AC_AB-1.png" alt="Venndir figure as in the previous figure, this time highlighting one overlap region between 'A - C' and 'A - B', with 214 shared hits which are all concordant in direction. The other set represents the contrast 'B - C', and therefore hits in the highlighted region can only be concordant in direction." width="960"></p>
<ul>
<li>
<p>These <strong>214</strong> hits are “distinctively A”, in that
they are present <code>"A - C"</code> and <code>"A - B"</code>. There
are two explanations for concordance or discordance here:</p>
<ul>
<li>Note these hits are not present in <code>"B - C"</code> therefore,
they could be considered “uniquely A”.</li>
<li>High concordance is expected, and suggests that these “A” changes
are unique to “A”, and have significantly stronger effect in “A” than in
“B”.</li>
<li>Discordance really should be extremely unlikely, since it would
require changes between <code>"A - B"</code> which are also not
<code>"B - C"</code> changes.</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">v4</span> <span class="op">&lt;-</span> <span class="fu">venndir</span><span class="fu">::</span><span class="fu"><a href="../reference/modify_venndir_overlap.html">highlight_venndir_overlap</a></span><span class="op">(</span><span class="va">v</span>,</span>
<span>      overlap_set<span class="op">=</span><span class="st">"B - C&amp;A - B"</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v4</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/venn_compare_BC_AB-1.png" alt="Venndir figure as in the previous figure, highlighting another overlap region between 'B - C' and 'A - B', with 126 hits which are all discordant in direction. Again, due to the organization of this three-way Venn diagram, hits in this overlap region can only be discordant in direction." width="960"></p>
<ul>
<li>These <strong>126</strong> hits are “distinctively B”, completely
analogous to the “distinctively A” example above.</li>
</ul>
<p>However, note that all hits are discordant. It may be confusing at
first, however, this effect is due to the order of the comparisons. “B”
is the test group in <code>"B - C"</code>, but “B” is the control group
in <code>"A - B"</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">v5</span> <span class="op">&lt;-</span> <span class="fu">venndir</span><span class="fu">::</span><span class="fu"><a href="../reference/modify_venndir_overlap.html">highlight_venndir_overlap</a></span><span class="op">(</span><span class="va">v</span>,</span>
<span>      overlap_set<span class="op">=</span><span class="st">"A - C&amp;B - C"</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v5</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/venn_compare_AC_BC-1.png" alt="Venndir figure as in the previous figure, showing the overlap between 'A - C' and 'B - C', with 726 genes which are all concordant in direction. In this overlap region, hits may be concordant or discordant, and the observation of entirely concordant hits is a finding." width="960"></p>
<ul>
<li>These <strong>726</strong> hits are shared by A and B compared with
Control. The key question is whether the hits are concordant. This
example represents the driving reason Venndir was created.</li>
</ul>
<p>There are two conceptual results:</p>
<ol style="list-style-type: decimal">
<li>
<p>Overlap</p>
<ul>
<li>High overlap suggests “the same components are involved”.</li>
<li>Low overlap suggests “not many of the same components are
involved.”</li>
</ul>
</li>
<li>
<p>Directional concordance</p>
<ol style="list-style-type: lower-alpha">
<li>
<strong>High concordance</strong> suggests the components are
affected in the same way. This outcome suggests ‘A’ and ‘B’ have similar
biological effect.</li>
<li>
<strong>Mixed concordance</strong> suggests that components are
<em>not</em> affected similarly. In fact, this outcome argues against
‘A’ and ‘B’ having similar biological effects.</li>
<li>
<strong>Low concordance</strong> suggests the components are
affected in the <em>opposite</em> way. (This outcome may be equally
compelling.) This outcome suggests ‘A’ and ‘B’ have opposite biological
effects, for example disease and treatment.</li>
</ol>
</li>
</ol>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">v6</span> <span class="op">&lt;-</span> <span class="fu">venndir</span><span class="fu">::</span><span class="fu"><a href="../reference/modify_venndir_overlap.html">highlight_venndir_overlap</a></span><span class="op">(</span><span class="va">v</span>,</span>
<span>      overlap_set<span class="op">=</span><span class="st">"A - C&amp;B - C&amp;A - B"</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v6</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/venn_compare_AC_BC_AC-1.png" alt="Venndir figure as in the previous figure, highlighting the three-way overlap between 'A - C', 'B - C', and 'A - B', and showing 13 shared hits in all three contrasts. This overlap region is not expected to have any particular concordance, and is difficult to interpret without reviewing the underlying data." width="960"></p>
<p>In my opinion, the concordance here is not useful to interpret. It
may be interesting to display <code>overlap_type="each"</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">veach</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">fit2decide1</span>,</span>
<span>      do_plot<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>      sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>      overlap_type<span class="op">=</span><span class="st">"each"</span><span class="op">)</span>;</span>
<span>   <span class="va">v6each</span> <span class="op">&lt;-</span> <span class="fu">venndir</span><span class="fu">::</span><span class="fu"><a href="../reference/modify_venndir_overlap.html">highlight_venndir_overlap</a></span><span class="op">(</span><span class="va">veach</span>,</span>
<span>      overlap_set<span class="op">=</span><span class="st">"A - C&amp;B - C&amp;A - B"</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">v6each</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>&lt;img
src=“/Users/jamesward/Projects/Ward/venndir/docs/articles/venndir_gene_expression_files/figure-html/venn_compare_AC_BC_AC_each-1.png”
alt=“Venndir figure as in the previous figure, highlighting the
three-way overlap between ‘A - C’, ‘B - C’, and ‘A - B’. This figure
uses ‘overlap_type=“each”’ which shows each combination of up and down.
All 13 hits are concordant in the first two contrasts, and only 3 hits
are discordant in the third contrast. The pattern up-up-down suggests
that the effects are shared in A and B compared to C, and that the
magnitude in B is significantly greated in B.” width=“960” /&gt;</p>
<ul>
<li>
<p>This plot is mainly useful when A and B have concordant effects
compared with Control.</p>
<ul>
<li>When A and B are concordant, results should typically begin
<code>^^</code> or <code>vv</code>.</li>
<li>The third arrow indicates the contrast with higher magnitude.</li>
<li>For example <code>^^^</code> is up in A-C, up in B-C, and higher in
A-B. Therefore A has a greater effect (up) than B.</li>
<li>For example <code>vvv</code> is down in A-C, down in B-C, and even
lower in A-B. Therefore A has a greater effect (down) than B.</li>
<li>However <code>^^v</code> is up in A-C, up in B-C, and lower in A-B.
Therefore B has a greater effect (up) than A.</li>
<li>Similarly <code>vv^</code> is down in A-C, down in B-C, and higher
in A-B. Therefore B has a greater effect (down) than A.</li>
</ul>
</li>
<li>
<p>When A and B are discordant, the third arrow should match the
first arrow, and is therefore not informative.</p>
<ul>
<li>
<code>^v^</code> is up in A-C, down in B-C. Therefore when comparing
‘A-B’ the direction must also be up.</li>
<li>
<code>v^v</code> is down in A-C, up in B-C. Therefore when comparing
‘A-B’ the direction must also be down.</li>
</ul>
</li>
<li>
<p>To state it clearly, there should never be a scenario with
<code>^vv</code> or <code>v^^</code> for the contrasts A-C, B-C, A-B. If
this pattern occurs, it suggests two possibilities:</p>
<ol style="list-style-type: decimal">
<li>The contrasts are reversed, for example <code>'B-A'</code> instead
of <code>'A-B'</code>. This is fine, but adjust the expectations
accordingly.</li>
<li>There may be high variability in the data, causing the estimated
directions to be unstable. It may occur when using imputed data in place
of missing or sparse data, or when there are outlier measurements which
are order of magnitude (10-fold or more) different than typical expected
results.</li>
</ol>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="venndir-with-deseq2-results">Venndir with DESeq2 results<a class="anchor" aria-label="anchor" href="#venndir-with-deseq2-results"></a>
</h2>
<p>The same data is processed using DESeq2, with the caveat that this is
Illumina microarray data - not RNA-seq data as intended for DESeq2. This
example is purely to demonstrate the mechanics of the process.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="co"># DESeq2 DataSet</span></span>
<span>   <span class="va">targets</span><span class="op">$</span><span class="va">DonorNum</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"E"</span>, <span class="va">targets</span><span class="op">$</span><span class="va">Donor</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="va">targets</span><span class="op">$</span><span class="va">CellType</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">targets</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span></span>
<span>   <span class="va">countData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">E</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>;</span>
<span>   <span class="va">countData</span><span class="op">[</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html" class="external-link">as.integer</a></span><span class="op">(</span><span class="va">countData</span><span class="op">)</span></span>
<span>   <span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>      <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></span><span class="op">(</span></span>
<span>         countData<span class="op">=</span><span class="va">countData</span>,</span>
<span>         rowData<span class="op">=</span><span class="va">y</span><span class="op">$</span><span class="va">genes</span>,</span>
<span>         colData<span class="op">=</span><span class="va">targets</span>,</span>
<span>         design<span class="op">=</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">CellType</span> <span class="op">+</span> <span class="va">DonorNum</span><span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>   <span class="co"># run DESeq</span></span>
<span>   <span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span>   <span class="co"># calculate contrasts</span></span>
<span>   <span class="va">use_contrasts</span> <span class="op">&lt;-</span> <span class="va">contrasts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">4</span><span class="op">)</span>, <span class="op">]</span>;</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">use_contrasts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CellType"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">contrasts</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">resultsNames</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span>   </span>
<span>   <span class="co"># iterate each contrast to produce DESeqResults</span></span>
<span>   <span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://jmw86069.github.io/jamba/reference/nameVector.html" class="external-link">nameVector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">use_contrasts</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span>,</span>
<span>         saveCols<span class="op">=</span><span class="st">"SYMBOL"</span>,</span>
<span>         contrast<span class="op">=</span><span class="va">use_contrasts</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>   <span class="op">}</span><span class="op">)</span></span>
<span>   <span class="co"># filter for statistical hits</span></span>
<span>   <span class="va">res_hits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">ires</span><span class="op">)</span><span class="op">{</span></span>
<span>      <span class="va">isub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ires</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>            <span class="va">padj</span> <span class="op">&lt;=</span> <span class="fl">0.01</span> <span class="op">&amp;</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>;</span>
<span>      <span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://jmw86069.github.io/jamba/reference/nameVector.html" class="external-link">nameVector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">isub</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">isub</span><span class="op">)</span><span class="op">)</span></span>
<span>   <span class="op">}</span><span class="op">)</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res_hits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res_hits</span><span class="op">)</span>, <span class="st">"\nDESeq2"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; -- note: fitType='parametric', but the dispersion trend was not well captured by the</span></span>
<span><span class="co">#&gt;    function: y = a/x + b, and a local regression fit was automatically substituted.</span></span>
<span><span class="co">#&gt;    specify fitType='local' or 'mean' to avoid this message next time.</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span></code></pre></div>
<div class="section level3">
<h3 id="deseq2-hits">DESeq2 hits<a class="anchor" aria-label="anchor" href="#deseq2-hits"></a>
</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_deseq2</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">res_hits</span>,</span>
<span>      proportional<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>      template<span class="op">=</span><span class="st">"tall"</span>,</span>
<span>      sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/deseq2_3a-1.png" alt="Venndir proportional Euler diagram using statistical hits from DESeq2 analysis. The vast majority of hits shared in these two contrasts are concordant, 1358 hits are up in both, 1360 hits are down in both, and only 25 hits have discordant direction." width="960"></p>
</div>
<div class="section level3">
<h3 id="compare-limma-to-deseq2">Compare limma to DESeq2<a class="anchor" aria-label="anchor" href="#compare-limma-to-deseq2"></a>
</h3>
<p>It’s only natural to want to compare limma and DESeq2… Remember that
DESeq2 is used here with microarray data only for educational purposes,
and is not generally recommended for array data otherwise. That said, it
does perform fairly well.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_deseq2</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="va">use_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="va">limmalist</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>      <span class="va">res_hits</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span>;</span>
<span>   <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">use_list</span>,</span>
<span>      font_cex<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>      overlap_type<span class="op">=</span><span class="st">"agreement"</span>,</span>
<span>      <span class="co"># alpha_by_counts=TRUE,</span></span>
<span>      template<span class="op">=</span><span class="st">"tall"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_deseq2_1-1.png" alt="Four-way Venndir diagram comparing limma results to DESeq2 results. Hits from the same contrast are shared at a high level across the two methods, as anticipated, in fact the region with the most probes is the four-way overlap, which suggests great agreement in the two contrasts, and the two methodologies." width="960"></p>
<p>Quick notes:</p>
<ul>
<li>Notice there are more similarities within the same contrast (1,223
and 827). These are likely to be robust hits, detected by both limma and
DESeq2.</li>
<li>There is a much lower overlap within each platform (519, 117) across
the two contrasts. These are likely to be specific to each method, owing
to limma’s specialization in microarrays.</li>
<li>There are no shared hits across methods for different contrasts,
which is also encouraging.</li>
<li>Almost all shared hits are concordant.</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
