<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Venndir with Gene Expression Data • venndir</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Venndir with Gene Expression Data">
<meta property="og:description" content="venndir">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">venndir</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.45.900</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/venndir_gene_expression.html">Venndir with Gene Expression Data</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://github.com/jmw86069/venndir/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Venndir with Gene Expression Data</h1>
            
      
      <small class="dont-index">Source: <a href="http://github.com/jmw86069/venndir/blob/HEAD/vignettes/venndir_gene_expression.Rmd" class="external-link"><code>vignettes/venndir_gene_expression.Rmd</code></a></small>
      <div class="hidden name"><code>venndir_gene_expression.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/jmw86069/venndir" class="external-link">venndir</a></span><span class="op">)</span></span></code></pre></div>
<p>Venndir was inspired by gene expression data, which contains up/down
directionality. Here are some examples showing how gene expression data
can be used in <code>venndir</code>.</p>
<div class="section level3">
<h3 id="generic-example">Generic example<a class="anchor" aria-label="anchor" href="#generic-example"></a>
</h3>
<p>The <code><a href="../reference/make_venn_test.html">venndir::make_venn_test()</a></code> function is useful to
generate test data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">setlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_venn_test.html">make_venn_test</a></span><span class="op">(</span><span class="fl">2500</span>, <span class="fl">3</span>,</span>
<span>  sizes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">400</span>, <span class="fl">500</span>, <span class="fl">200</span><span class="op">)</span>,</span>
<span>  do_signed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">setlist</span>, overlap_type<span class="op">=</span><span class="st">"agreement"</span><span class="op">)</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/generic_1-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="venndir-with-limma-results">Venndir with limma results<a class="anchor" aria-label="anchor" href="#venndir-with-limma-results"></a>
</h3>
<p>The <code>limma</code> package provides an extensive set of gene
expression data analysis tools. The steps below reproduce the
<code>Limma Users Guide</code> example <code>"Illumina Use Case"</code>
in Section 17.3. This example requires data available at <a href="http://bioinf.wehi.edu.au/marray/IlluminaCaseStudy" class="external-link uri">http://bioinf.wehi.edu.au/marray/IlluminaCaseStudy</a>
copied to the current working directory.</p>
<p>For the sake of this example, the steps assume the data is located in
a folder <code>[HOME]/IlluminaCaseStudy</code>, where
<code>[HOME]</code> is the user home directory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">illuminadir</span> <span class="op">&lt;-</span> <span class="st">"~/IlluminaCaseStudy"</span>;</span>
<span><span class="va">do_limma</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>;</span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/limma" class="external-link">limma</a></span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="va">illuminadir</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">do_limma</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span>;</span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">targets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/readTargets.html" class="external-link">readTargets</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">illuminadir</span><span class="op">)</span></span>
<span>  <span class="co"># import Illumina data</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/read.ilmn.html" class="external-link">read.ilmn</a></span><span class="op">(</span>files<span class="op">=</span><span class="st">"probe profile.txt.gz"</span>,</span>
<span>     ctrlfiles<span class="op">=</span><span class="st">"control probe profile.txt.gz"</span>,</span>
<span>     other.columns<span class="op">=</span><span class="st">"Detection"</span>,</span>
<span>    path<span class="op">=</span><span class="va">illuminadir</span><span class="op">)</span></span>
<span>  <span class="co"># define expressed probes</span></span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/nec.html" class="external-link">neqc</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">expressed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">other</span><span class="op">$</span><span class="va">Detection</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span>;</span>
<span>  <span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">expressed</span>,<span class="op">]</span></span>
<span>  <span class="co"># calculate within-patient correlations</span></span>
<span>  <span class="va">ct</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">targets</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span></span>
<span>  <span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span><span class="op">+</span><span class="va">ct</span><span class="op">)</span>;</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">ct</span><span class="op">)</span>;</span>
<span>  <span class="va">dupcor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">duplicateCorrelation</a></span><span class="op">(</span><span class="va">y</span>,</span>
<span>     <span class="va">design</span>,</span>
<span>     block<span class="op">=</span><span class="va">targets</span><span class="op">$</span><span class="va">Donor</span><span class="op">)</span>;</span>
<span>  <span class="va">dupcor</span><span class="op">$</span><span class="va">consensus.correlation</span>;</span>
<span></span>
<span>  <span class="co"># fit linear model, paired by patient, with correlations</span></span>
<span>  <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">y</span>,</span>
<span>     <span class="va">design</span>,</span>
<span>     block<span class="op">=</span><span class="va">targets</span><span class="op">$</span><span class="va">Donor</span>,</span>
<span>     correlation<span class="op">=</span><span class="va">dupcor</span><span class="op">$</span><span class="va">consensus.correlation</span><span class="op">)</span>;</span>
<span>  <span class="co"># define contrasts</span></span>
<span>  <span class="va">contrasts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span><span class="va">ML</span><span class="op">-</span><span class="va">MS</span>, </span>
<span>     <span class="va">LP</span><span class="op">-</span><span class="va">MS</span>, </span>
<span>     <span class="va">ML</span><span class="op">-</span><span class="va">LP</span>, </span>
<span>     levels<span class="op">=</span><span class="va">design</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="co"># fit contrasts</span></span>
<span>  <span class="va">fit2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html" class="external-link">contrasts.fit</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">contrasts</span><span class="op">)</span>;</span>
<span>  </span>
<span>  <span class="co"># eBayes</span></span>
<span>  <span class="va">fit2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fit2</span>, trend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; Reading file ~/IlluminaCaseStudy/probe profile.txt.gz ... ...</span></span>
<span><span class="co">#&gt; Reading file ~/IlluminaCaseStudy/control probe profile.txt.gz ... ...</span></span></code></pre></div>
<p>The steps are straightforward except that this example uses a paired
linear model. Nonetheless the results is the same, the resulting object
is class <code>MArrayLM</code>.</p>
<p>The function <code><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">limma::decideTests()</a></code> applied adjusted
P-value and log2 fold change filtering, and produces class
<code>TestResults</code> which is actually a signed incidence matrix.
Take a look at the output:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># decideTests() creates a signed incidence matrix</span></span>
<span>  <span class="va">fit2decide</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">fit2</span>, method<span class="op">=</span><span class="st">"global"</span><span class="op">)</span>;</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fit2decide</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; TestResults matrix</span></span>
<span><span class="co">#&gt;               Contrasts</span></span>
<span><span class="co">#&gt;                ML - MS LP - MS ML - LP</span></span>
<span><span class="co">#&gt;   ILMN_2055271       0       0       0</span></span>
<span><span class="co">#&gt;   ILMN_1653355       0       0       0</span></span>
<span><span class="co">#&gt;   ILMN_1787689       0       0       0</span></span>
<span><span class="co">#&gt;   ILMN_1745607      -1      -1       0</span></span>
<span><span class="co">#&gt;   ILMN_1735045       0       0       0</span></span>
<span><span class="co">#&gt;   ILMN_1659452       0       0       0</span></span></code></pre></div>
<p>A signed incidence matrix can be converted to <code>setlist</code>
with <code><a href="../reference/im_value2list.html">im_value2list()</a></code>. Then call <code><a href="../reference/venndir.html">venndir()</a></code>.</p>
<p>The <code>"im_value"</code> refers to an incidence matrix with values
(signs), and <code>"2list"</code> will convert to a <code>list</code>.
The <code>list</code> returned will contain named vectors, where names
are the item labels, and values are the signs.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">limmalist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide</span><span class="op">)</span>;</span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_3-1.png" width="960"></p>
<p>A few features are noticeable:</p>
<ol style="list-style-type: decimal">
<li><p>The <code>limma</code> contrasts <code>"ML - MS"</code> and
<code>"LP - MS"</code> are shown in each Venn circle. The numbers in
each circle represent statistically significant changes given the
<code>limma</code> thresholds.</p></li>
<li><p>There is fairly high overlap between these gene lists,
<code>3,681</code> probes are shared, which is more than not shared
<code>2,922</code> and <code>2,303</code>, respectively.</p></li>
<li>
<p>Almost all the overlapping probes are changing in the same
direction in these two contrasts</p>
<ul>
<li>
<code>1,800</code> probes are up in both contrasts</li>
<li>
<code>1,856</code> probes are down in both contrasts</li>
<li>
<code>25</code> probes disagree in up-down or down-up direction</li>
</ul>
</li>
<li><p>Implied in #3, the shared probes are roughly evenly distributed
between up and down.</p></li>
</ol>
<div class="section level4">
<h4 id="proportional-venn-diagram-euler">Proportional Venn diagram (Euler)<a class="anchor" aria-label="anchor" href="#proportional-venn-diagram-euler"></a>
</h4>
<p>The data can be drawn with proportional circles, otherwise known as a
Euler diagram, or proportional Venn diagram.</p>
<p>For this diagram, we will plot <code>sets=c(1, 3)</code> mostly
because these are more visually interesting.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">limmalist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide</span><span class="op">)</span>;</span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, proportional<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_3p-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="overlap_type-agreement">overlap_type “agreement”<a class="anchor" aria-label="anchor" href="#overlap_type-agreement"></a>
</h4>
<p>An alternative approach to represent concordance is by “agreement”,
which combines up-up and down-down into one summary number.</p>
<p>Use the argument <code>overlap_type="agreement"</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">limmalist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide</span><span class="op">)</span>;</span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, overlap_type<span class="op">=</span><span class="st">"agreement"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_4-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="overlap_type-each">overlap_type “each”<a class="anchor" aria-label="anchor" href="#overlap_type-each"></a>
</h4>
<p>Finally, you can represent all the directional changes using
<code>overlap_type="each"</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">limmalist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide</span><span class="op">)</span>;</span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, overlap_type<span class="op">=</span><span class="st">"each"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_5-1.png" width="960"></p>
</div>
<div class="section level4">
<h4 id="display-item-labels">Display item labels<a class="anchor" aria-label="anchor" href="#display-item-labels"></a>
</h4>
<p>It is sometimes interesting to show item labels inside the Venn
diagram – but 3,600 text labels are too many to be displayed!</p>
<p>For the purpose of this example, we will filter statistical hits
using a higher fold change, and more significant P-value threshold –
just to reduce the labels.</p>
<p>In practice, examples with &gt;3000 hits should probably never be
labeled unless printing full size on a poster (and even then, with small
font!)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># increase stringency of statistical filtering</span></span>
<span>  <span class="va">fit2decide2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">fit2</span>,</span>
<span>    p.value<span class="op">=</span><span class="fl">1e-4</span>,</span>
<span>    lfc<span class="op">=</span><span class="fl">4</span><span class="op">)</span>;</span>
<span>  <span class="co"># convert to signed list</span></span>
<span>  <span class="va">limmalist2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide2</span><span class="op">)</span>;</span>
<span>  <span class="va">limmalist2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">limmalist2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/jamba/man/nameVector.html" class="external-link">nameVector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="co"># display item labels</span></span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist2</span>,</span>
<span>    sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>    poly_alpha<span class="op">=</span><span class="fl">0.3</span>,</span>
<span>    show_labels<span class="op">=</span><span class="st">"Ni"</span>,</span>
<span>    show_items<span class="op">=</span><span class="st">"sign item"</span>,</span>
<span>    item_cex<span class="op">=</span><span class="fl">1</span>,</span>
<span>    item_degrees<span class="op">=</span><span class="fl">4</span>,</span>
<span>    max_items<span class="op">=</span><span class="fl">1000</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_6-1.png" width="960"></p>
</div>
</div>
<div class="section level3">
<h3 id="display-item-directions">DIsplay item directions<a class="anchor" aria-label="anchor" href="#display-item-directions"></a>
</h3>
<p>The display of item labels brings up some potential benefits: it does
indicate the relative density of labels, and relative quantity of
up/down/mixed direction.</p>
<p>You can customize the item label so that it only displays the
directional sign, and not the item. This technique may be good for
higher number of items.</p>
<p>Use argument <code>show_items="sign"</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">vo6s</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist2</span>,</span>
<span>    sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>    poly_alpha<span class="op">=</span><span class="fl">0.3</span>,</span>
<span>    show_labels<span class="op">=</span><span class="st">"Ni"</span>,</span>
<span>    show_items<span class="op">=</span><span class="st">"sign"</span>,</span>
<span>    item_cex<span class="op">=</span><span class="fl">1</span>,</span>
<span>    max_items<span class="op">=</span><span class="fl">1000</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_6s-1.png" width="960"></p>
<div class="section level4">
<h4 id="convert-probes-to-genes">Convert probes to genes<a class="anchor" aria-label="anchor" href="#convert-probes-to-genes"></a>
</h4>
<p>The first item label figure highlights an important issue:
<code>limma</code> and similar analysis tools test the probes, not the
genes.Therefore the results represent Illumina probes, which are not
very helpful by themselves.</p>
<p>Fortunately in this case, the gene data is available in
<code>y$genes</code>, so we can convert each row to a gene symbol.
However there are multiple probes per gene symbol. For the sake of this
example, we will choose the first of each gene symbol from the
statistical hits.</p>
<blockquote>
<p>There is another Jam R package function that may be helpful:
<code><a href="https://rdrr.io/pkg/genejam/man/freshenGenes.html" class="external-link">genejam::freshenGenes()</a></code>. This function takes one of more
columns of gene symbols, gene accessions, gene identifiers, and returns
the “best matching result” using Bioconductor gene annotation data.</p>
</blockquote>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># convert probe hits to gene hits</span></span>
<span>  <span class="va">limmalist2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/im_value2list.html">im_value2list</a></span><span class="op">(</span><span class="va">fit2decide2</span><span class="op">)</span>;</span>
<span>  <span class="co"># iterate the hit list to convert to gene</span></span>
<span>  <span class="va">limmalist2g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">limmalist2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">kdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>check.names<span class="op">=</span><span class="cn">FALSE</span>, <span class="va">i</span><span class="op">)</span>;</span>
<span>    <span class="co"># use gene symbol</span></span>
<span>    <span class="va">kdf</span><span class="op">$</span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">kdf</span><span class="op">)</span>, <span class="st">"SYMBOL"</span><span class="op">]</span>;</span>
<span>    <span class="co"># subset for unique genes with non-empty value</span></span>
<span>    <span class="va">kdf_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">kdf</span>, <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># make a vector of signed direction</span></span>
<span>    <span class="va">kdf_genes</span> <span class="op">&lt;-</span> <span class="va">kdf_sub</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>;</span>
<span>    <span class="co"># name the vector using gene symbol</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">kdf_genes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">kdf_sub</span><span class="op">$</span><span class="va">genes</span>;</span>
<span>    <span class="va">kdf_genes</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist2g</span>, sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>    show_labels<span class="op">=</span><span class="st">"Ni"</span>,</span>
<span>    show_items<span class="op">=</span><span class="st">"sign item"</span>,</span>
<span>    poly_alpha<span class="op">=</span><span class="fl">0.3</span>,</span>
<span>    item_cex<span class="op">=</span><span class="fl">1</span>,</span>
<span>    max_items<span class="op">=</span><span class="fl">1000</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_7-1.png" width="960"></p>
</div>
</div>
<div class="section level3">
<h3 id="three-way-directional-venn">Three-way directional Venn<a class="anchor" aria-label="anchor" href="#three-way-directional-venn"></a>
</h3>
<p>The examples above purposefully used only two contrasts, because
those two contrasts show very high concordance. An experiment design
with three groups is quite common, and adding directionality for the
third contrast can be confusing, and still useful. (See section below on
interpreting three-group Venn diagrams.)</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">do_limma</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/venndir.html">venndir</a></span><span class="op">(</span><span class="va">limmalist</span>,</span>
<span>    sets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    overlap_type<span class="op">=</span><span class="st">"agreement"</span><span class="op">)</span>;</span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="venndir_gene_expression_files/figure-html/limma_8-1.png" width="960"></p>
<div class="section level4">
<h4 id="comments-on-converting-probe-to-gene">Comments on converting probe to gene<a class="anchor" aria-label="anchor" href="#comments-on-converting-probe-to-gene"></a>
</h4>
<p>When converting probes to genes, I usually run a quick test to see if
any genes have statistically significant probes that are “up” and “down”
– I call these “bi-directional genes”. If there are no bi-directional
genes, then choosing one entry per gene is reasonable. I leave this
evaluation to the scientist, but please post an issue if you have
specific questions.</p>
<p>For RNA-seq data, the input data matrix may already contain gene
expression values – as when using <code><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport::tximport()</a></code> with
the argument <code>tx2gene</code>; or when importing
<code>featureCounts</code> data where each row represents one gene
identifier. In those cases, no conversion is required.</p>
</div>
</div>
<div class="section level3">
<h3 id="interesting-examples-and-how-to-interpret-them">Interesting examples and how to interpret them<a class="anchor" aria-label="anchor" href="#interesting-examples-and-how-to-interpret-them"></a>
</h3>
<div class="section level4">
<h4 id="three-group-analysis">Three-group analysis<a class="anchor" aria-label="anchor" href="#three-group-analysis"></a>
</h4>
<p>One common type of analysis involves three experimental groups. For
example:</p>
<ol style="list-style-type: decimal">
<li>Control group</li>
<li>Treatment A</li>
<li>Treatment B</li>
</ol>
<p>The typical contrasts are:</p>
<ol style="list-style-type: decimal">
<li>Treatment A - Control</li>
<li>Treatment B - Control</li>
<li>Treatment B - Treatment A</li>
</ol>
<p>Nothing particularly unusual about the experiment design, nor the
contrasts, these are fairly standard and straightforward.</p>
<p>When creating a three-way Venn diagram, the results are also fairly
straightforward. When the directionality is included, it can be more
confusing, but also potentially much more informative.</p>
<p>There are a few potential scenarios:</p>
<ol style="list-style-type: decimal">
<li>Treatment A does not affect similar genes as does Treatment B. This
result is seen with low overlaps.</li>
<li>Treatment A affects similar genes as Treatment B, but not with any
particular concordance.</li>
<li>Treatment A affects similar genes as Treatment B, with extremely
high concordance.</li>
</ol>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
